symengine/parser/tokenizer.re

```
@@ -0,0 +1,176 @@
+    unsigned char *tok;
+    bool eof;
+
+    std::istream &file;
+
+    SymEngine::ParserBase::STYPE__ *val;
+
+    input_t(std::istream &f)
+        : buf()
+        , lim(buf + SIZE)
+        , cur(lim)
+        , mar(lim)
+        , tok(lim)
+        , eof(false)
+        , file(f)
+    {}
+    bool fill(size_t need)
+    {
+        if (eof) {
+            return false;
+        }
+        const size_t free = tok - buf;
+        if (free < need) {
+            return false;
+        }
+        memmove(buf, tok, lim - tok);
+        lim -= free;
+        cur -= free;
+        mar -= free;
+        tok -= free;
+        lim += sread((char*)lim, 1, free, file);
+        if (lim < buf + SIZE) {
+            eof = true;
+            memset(lim, 0, YYMAXFILL);
+            lim += YYMAXFILL;
+        }
+        return true;
+    }
+};
+
+int yylex(input_t &in)
+{
+    for (;;) {
+        in.tok = in.cur;
+        /*!re2c
+            re2c:define:YYCURSOR = in.cur;
+            re2c:define:YYMARKER = in.mar;
+            re2c:define:YYLIMIT = in.lim;
+            re2c:yyfill:enable = 1;
+            re2c:define:YYFILL = "if (!in.fill(@@)) return 0;";
+            re2c:define:YYFILL:naked = 1;
+            re2c:define:YYCTYPE = "unsigned char";
+
+            end = "\x00";
+
+
+            whitespace = [ \t\v\n\r]+;
+            dig = [0-9];
+            char =  [\x80-\xff] | [a-zA-Z_];
+            operators = "-"|"+"|"/"|"("|")"|"*"|","|"^"|"~"|"<"|">"|"&"|"|";
+
+            pows = "**"|"@";
+            le = "<=";
+            ge = ">=";
+            eqs = "==";
+            ident = char (char | dig)*;
+            numeric = (dig*"."?dig+([eE][-+]?dig+)?) | (dig+".");
+            implicitmul = numeric ident;
+
+            *   { printf("ERR unknown token\n"); return 0; }
+            end {
+                    if (in.lim - in.tok == YYMAXFILL) {
+                        return 0;
+                    } else {
+                        printf("ERR NULL.\n");
```

These two print statements are the only ugly parts of the code. Essentially they signal an error. I think we should raise an exception instead?

---

Yes, a `SymEngine::ParseError`.

---

My understanding that both of these error conditions will disappear after I rework the tokenizer per my comment https://github.com/symengine/symengine/pull/1586#issuecomment-517745895.